### DISCUSSION TITLE
Accelerator/shortcut keys

### DISCUSSION THREAD

**Discussion by ghost** [keepassxreboot/keepassxc#9068] [Ideas]
I'm a big fan of KeepassXC and think it is generally superior to Keepass, but there are couple of issues that keep arising in my use.

I use Nuance's Dragon Professional speech recognition, and Keepass is not very Dragon-friendly. The first issue is that there are no accelerator keys in the Add Entry and Edit Entry windows, so it is difficult to navigate by voice and impossible to write custom commands to navigate.

There is also the issue that Dragon's Select-and-Say capability doesn't work in KeepassXC, unlike Keepass. That is probably a much more difficult issue to correct, unfortunately. Nuance hasn't given much guidance on how to make applications more friendly for Dragon. 

Thanks for considering these points.

**Comment by droidmonkey** [comment]
What is an accelerator key?

**Comment by ghost** [comment]
Keyboard shortcut, like Alt+N.  Keepass XC has them in the main window, such as Ctrl+N for new entry, but not in the Add Entry or Edit Entry windows, or so it seems to me.

 

Thanks for the response!

**Reply by droidmonkey** [reply]
But what would they be used for in the add/edit entry view? We do support most of the main window shortcuts when editing an entry.

**Comment by ghost** [comment]
To move between fields.  Say all I want to do is add a note to an existing entry.  Is there a way to get there without using a mouse?

**Reply by droidmonkey** [reply]
Press Tab

**Comment by ghost** [comment]
Yes.  But what if then I decide I want to go to another field, such as password?  I know, I know, press Alt-Tab or Tab as many times as it takes.  But it would be far easier if I could reliably navigate to at least one field using a keyboard shortcut.

**Reply by droidmonkey** [reply]
Do programs actually implement what you are asking for? I've never seen that ever

**Comment by michaelk83** [comment]
For accessiblity's sake, screen readers should be able to identify the label for each field, right? Assuming that's already the case, then Dragon could, in principle, navigate to a field by it associated label. No need for accelerators. If Dragon doesn't support that, that should be a feature request for them.

**Comment by ghost** [comment]
Yes, most do, in my experience. Will get back to you later today with examples. On Feb 5, 2023, at 1:41 PM, Jonathan White ***@***.***> wrote:﻿
Do programs actually implement what you are asking for? I've never seen that ever

**Comment by ghost** [comment]
Here is an example of a program (Keepass) with shortcut keys in the Edit Entry window:

 



 

Here's a good piece on the need for keyboard shortcuts: https://medium.com/uxreactor/the-ux-case-for-integrating-accelerators-in-your-designs-23825b212332

 

KeepassXC already provides keyboard shortcuts in the main window, just not in subsidiary windows.

 

Dragon doesn't have a screen reader function. It simply takes your spoken words and translates them into written words or custom commands. So there's no way for Dragon add navigation by associated label to its functionality.

 

Thanks again for considering this.

**Reply by droidmonkey** [reply]
Could you not just say: press tab 6 times?

We are not going to add these accelerators, they are undiscoverable and extremely niche. I'm sure they will add a lot of code / configuration bloat as well. It is likely that keepass did not implement them either, it might be a native dotNET offering.

**Comment by ghost** [comment]
Here’s an excerpt from a book written by a friend, Alan Cantor, who is an expert in assistive technologies:

 

Who needs to master the Windows® keyboard interface?

People need choices in how they operate a PC. A well-designed computer system allows users to interact with software in different ways. Ideally, the mouse and keyboard interfaces should be equally easy to learn and use. Unfortunately, operating Windows® by keyboard is less straightforward than by pointing-and-clicking.

Nevertheless, some people have no choice but to learn mouseless techniques. For example: 

*	Individuals who are totally blind cannot see the mouse pointer, and therefore cannot use a mouse. Most blind individuals use a standard keyboard in conjunction with “screen reader” software. A screen reader translates information into synthetic speech, and can handle any information that can be represented as text, including menus, word processing documents, and hypertext links on web pages.
*	People who are deaf and blind cannot use a mouse, either. Most deaf and blind individuals use a standard keyboard, but “read” the screen using a hardware device called a refreshable Braille display.
*	Individuals who have low-vision may be able to discern an enlarged mouse pointer, but finding and identifying targets may be visually taxing. In many situations, pressing two or three keys is easier and more reliable than reaching for the mouse, finding the pointer, hunting for a small screen object, and clicking on it. Using the keyboard instead of the mouse reduces or eliminates eye strain associated with pointing-and-clicking.
*	Users with upper-body mobility impairments, including one-handed typists, toe-typists, and people with dystonia and cerebral palsy, may also find it awkward to point-and-click. The same is true for individuals who access PCs with a mouthstick, headstick, or similar appliance.
*	Pointing and clicking may be difficult for people with musculoskeletal injuries. Pascarelli and Quilter estimate that 25% to 30% of computer users experience symptoms related to overusing their computers. For people who develop mouse-induced repetitive strain injuries, the ability to access a computer by keyboard can be a lifesaver.
*	People who have learning disabilities that affect hand-eye coordination may also find it easier to perform certain tasks by keyboard. In fact, some screen objects are so small that people who do not have hand-eye coordination problems find it difficult to target and click on them. Mouseless techniques can be a boon to these people.
*	When users develop temporary disabilities or are injured — for example, a broken arm, dislocated shoulder, or tennis elbow — the ability to use the mouse may be compromised, and a few keyboard techniques can be beneficial.
*	Older users and children sometimes have difficulties controlling a mouse for certain tasks. Double-clicking is especially difficult for some individuals. Knowledge of a few keyboard techniques can make it much easier for seniors and younger users to operate computers.

 

Finally, anyone who wants to work quickly and with fewer errors can benefit from knowing a few keyboard “tricks.” Many tasks that require precise mouse movements and accurate clicking can be done by pressing one or two keys on the keyboard. For reasons of efficiency, speed, and directness, power users often prefer using the keyboard to the mouse.

**Comment by ghost** [comment]
No, if you are in the Notes field, you can't even escape the field pressing Tab. I don't think there is any way to get out of the Notes field except by using a mouse.

 

Obviously, it's your choice whether or not to add these accelerators. Unfortunately, not including them makes KeePassXC less accessible for persons with disabilities, among others. For that reason, I do not think that keyboard accelerators are "niche".

 

Luckily, Keepass is a good alternative for such people.

 

Thanks again for considering this so quickly.

 

Matt Chambers

**Comment by droidmonkey** [comment]
We are going to fix the focus trap in notes: #9060

OK now I know what you are talking about, I went into KeePass and pressed Alt to see the underlines. In Qt those are called Buddies, I am experimenting with implementing this now.

I just tested it an the feature is rather annoying but works. When you press Alt the top menu is selected instead of showing the underlines. Also some of the obvious choices are already taken by the top menu so you end up toggling between the menu and the edit dialog. (Alt+T opens the Tools menu, but can also be used for Tags).

**Comment by ghost** [comment]
Thank you. The accessibility problem is still significant enough that I think I will simply use Keepass.

 

I do appreciate that you looked into my inquiry very quickly.
